cmd/pop/main.go

```diff
@@ -0,0 +1,435 @@
+		return err
+	}
+
+	msgDir := filepath.Dir(args.messagesFile)
+	failedFile := filepath.Join(msgDir, "failed_downloading")
+	failedJSON, err := json.MarshalIndent(&failed, "", "    ")
+	if err != nil {
+		return err
+	}
+
+	// This is not a failure condition, since the failed ones may be retried
+	_ = os.WriteFile(failedFile, failedJSON, 0o600)
+
+	fmt.Println(string(s))
+	return nil
+}
+
+func runUpload(args uploadArgs) error {
+	if args.messagesFile == "" {
+		return fmt.Errorf("you must provide a messages file")
+	}
+
+	if args.signedDir == "" {
+		return fmt.Errorf("you must provide a directory for the signed packages")
+	}
+
+	ctx := context.Background()
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return err
+	}
+
+	msgBytes, err := os.ReadFile(args.messagesFile)
+	if err != nil {
+		return err
+	}
+
+	msgs := []QueueMessageDeserialize{}
+	if err := json.Unmarshal(msgBytes, &msgs); err != nil {
+		return err
+	}
+
+	var errs error
+	nameToMsg := make(map[string][]QueueMessageDeserialize)
+	for _, msg := range msgs {
+		nameToMsg[msg.Content.Artifact.Name] = append(nameToMsg[msg.Content.Artifact.Name], msg)
+	}
+
+	failed := []QueueMessageDeserialize{}
+	successful := []QueueMessageDeserialize{}
+
+	fail := func(e error, f ...QueueMessageDeserialize) {
+		for _, c := range f {
+			failed = append(failed, c)
+		}
+		errs = errors.Join(errs, err)
+	}
+
+	client, err := azblob.NewClient(prodAccountName, cred, nil)
+	if err != nil {
+		return err
+	}
+
+outer:
+	for pkgBasename, messages := range nameToMsg {
+		signedPkgFilename := filepath.Join(args.signedDir, pkgBasename)
+		if _, err := os.Stat(signedPkgFilename); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("filename not found in the signing directory; signing likely failed for '%s'", pkgBasename))
+			continue
+		}
+
+		if len(messages) > 1 {
+			for i := 1; i < len(messages); i++ {
+				if messages[i-1].Content.Artifact.Sha256Sum != messages[i].Content.Artifact.Sha256Sum {
+					fail(
+						fmt.Errorf(
+							"messages encountered with same filename and different sha256 digests; manual intervention will be required. "+
+								"digest[%d]: %s digest[%d]: %s",
+							i-1, messages[i-1].Content.Artifact.Sha256Sum, i, messages[i].Content.Artifact.Sha256Sum),
+						messages...,
+					)
+					continue outer
+				}
+			}
+		}
+
+		msg := messages[0]
+
+		pkg := msg.Content.Spec.Pkg
+		version := fmt.Sprintf("%s+azure", msg.Content.Spec.Tag)
+		distro := msg.Content.Spec.Distro
+		pkgOS := msg.Content.Spec.OS()
+		sanitizedArch := strings.ReplaceAll(msg.Content.Spec.Arch, "/", "")
+
+		storagePath := fmt.Sprintf("%s/%s/%s/%s_%s/%s", pkg, version, distro, pkgOS, sanitizedArch, pkgBasename)
+		fmt.Fprintln(os.Stderr, storagePath)
+
+		f, err := os.Open(signedPkgFilename)
+		if err != nil {
+			fail(err, msg)
+			continue
+		}
+
+		if _, err := client.UploadFile(ctx, prodContainerName, storagePath, f, &azblob.UploadFileOptions{}); err != nil {
```

Comment:
We need to attach the sha256 digest as metadata to the blob.
Metadata key name should be `sha256` + the raw digest.
This is used by the PMC publish script to validate if a given package is already in PMC.
Looks like this can go into the `UploadFileOptions`.

Reply:
done